txtorcon/socks.py

```diff
@@ -579,7 +581,90 @@ def buildProtocol(self, addr):
 
 
 class SocksError(Exception):
-    pass
+
+    def __init__(self, message, errno=None):
+        super(SocksError, self).__init__(message)
+        self.errno = errno
+
+
+class SucceededReply(SocksError):
+
+    def __init__(self):
+        super(SucceededReply, self).__init__(
+            message='succeeded',
+            errno=0x00)
+
+
+class GeneralServerFailureError(SocksError):
+
+    def __init__(self):
+        super(GeneralServerFailureError, self).__init__(
+            message='general SOCKS server failure',
+            errno=0x01)
+
+
+class ConnectionNotAllowedError(SocksError):
+
+    def __init__(self):
+        super(ConnectionNotAllowedError, self).__init__(
+            message='connection not allowed by ruleset',
+            errno=0x02)
+
+
+class NetworkUnreachableError(SocksError):
+
+    def __init__(self):
+        super(NetworkUnreachableError, self).__init__(
+            message='Network unreachable',
+            errno=0x03)
+
+
+class HostUnreachableError(SocksError):
+
+    def __init__(self):
+        super(HostUnreachableError, self).__init__(
+            message='Host unreachable',
+            errno=0x04)
+
+
+class ConnectionRefusedError(SocksError):
+
+    def __init__(self):
+        super(ConnectionRefusedError, self).__init__(
+            message='Connection refused',
+            errno=0x05)
+
+
+class TtlExpiredError(SocksError):
+
+    def __init__(self):
+        super(TtlExpiredError, self).__init__(
+            message='TTL expired',
+            errno=0x06)
+
+
+class CommandNotSupportedError(SocksError):
+
+    def __init__(self):
+        super(CommandNotSupportedError, self).__init__(
+            message='Command not supported',
+            errno=0x07)
+
+
+class AddressTypeNotSupportedError(SocksError):
+
+    def __init__(self):
+        super(AddressTypeNotSupportedError, self).__init__(
+            message='Address type not supported',
+            errno=0x08)
+
+
+class SocksErrorFactory(object):
+    socks_errors = {cls().errno: cls for cls in SocksError.__subclasses__()}
+
+    @classmethod
+    def create(cls, errno):
+        return cls.socks_errors[errno]()
 
```

Comment:
Since this is just a single classmethod, I would make it a simple factory-method; this make extra sense with the above changes.

Also I would make it "private" since nothing should be creating errors besides code in here:

```
def _create_socks_error(code):
    try:
        return _socks_errors[code]()
    except KeyError:
        return SocksError(code=code, message="Unknown SOCKS error-code {}".format(code))
```

Well, that exact code above would need a little more smarts in SocksError's ``__init__``

I guess you can also see I think I prefer "code" or similar to "errno" (so we aren't "conflicting" with the popular C global variable name). Also, I prefer catching KeyError over "check, then access" when doing dict stuff (there's another one you did with the "check-first" way above).